`ot_i2c_host_proxy`: implement device for external I2C communication

ziuziakowska opened the pull request:
This is an I2C device that allows external software such as `opentitanlib` to communicate with and issue commands to devices on a QEMU I2C bus. The communication is done through a `chardev`.

See `docs/i2c_host_proxy.md` for usage and protocol information.

The device can be instantiated with these command line parameters:

`-chardev pty,id=i2cN -device ot-i2c-transport,bus=<bus>,chardev=i2cN`, to instantiate on an I2C bus.

See also https://github.com/lowRISC/opentitan/pull/28450 for an implementation and test harness using this device.

ziuziakowska committed:
[ot] hw/opentitan: ot_i2c_host_proxy: I2C Host Proxy implementation

This device is a I2C host device that is externally driven over a
chardev interface. It can be used to issue read and write transfers with
devices on the QEMU I2C buses from outside.

This will primarily be used by `opentitanlib` for its QEMU "transport"
implementation, which will issue transfers primarily with the OpenTitan
I2C target block. However, all devices on the I2C bus are accessible,
and any driver that speaks the wire protocol can be used too.

This device is modelled as a I2C bus device, but it cannot be interacted
with by other bus devices and cannot have transactions targetting it.

Signed-off-by: Alice Ziuziakowska <a.ziuziakowska@lowrisc.org>

ziuziakowska committed:
[ot] hw: opentitan: add I2C host proxy device to Kconfig and meson

Signed-off-by: Alice Ziuziakowska <a.ziuziakowska@lowrisc.org>

rivos-eblot commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,312 @@
+#include "hw/qdev-properties.h"
+#include "trace.h"
+
+#define TYPE_OT_I2C_TRANSPORT "ot-i2c-transport"
```
This is probably part of a .h header file.

rivos-eblot commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,312 @@
+
+#define CMD_FIFO_SIZE 256u
+
+enum i2c_cmd_parser_state {
```
Beware of QEMU syntax, should usually by a CamelCased version. As the enum itself is used to declare a variable of this type, you also need an (anonymous) typedef.

rivos-eblot commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,312 @@
+    bool is_read_command;
+    uint8_t byte;
+
+    I2CBus *bus;
```
As this is not a property, it would better be moved above (before cmd_bytes_fifo for ex.)

ziuziakowska replied:
I think the bus is a property inherited from `I2CSlave`, but I will move it if only the new properties should be at the bottom (in this case only `chardev`)

rivos-eblot commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,312 @@
+
+static void ot_i2c_transport_command_byte(OtI2CTransportState *s, uint8_t byte)
+{
+    int nibble;
```
prefer the original type, even if they are eq: `gint`

rivos-eblot commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,312 @@
+            s->parser_state = OT_I2C_CMD_INIT;
+            break;
+        }
+        s->byte = (s->byte << 4u) | nibble;
```
beware of mixing type, you likely want an explicit cast (from `gint`) to `uint8_t`

rivos-eblot commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,312 @@
+            s->is_read_command ? OT_I2C_CMD_BYTE : OT_I2C_CMD_PAYLOAD_1;
+        break;
+    case OT_I2C_CMD_PAYLOAD_1:
+        // parse first nibble of write payload byte
```
Prefer QEMU "C" comment styles (see checkpatch tool)

rivos-eblot commented on hw/opentitan/ot_i2c_host_proxy.c:
```diff
@@ -0,0 +1,312 @@
+{
+    (void)data;
+
+    DeviceClass *dc = DEVICE_CLASS(klass);
```
You probably need reset handler(s) to reinitialize your state machine, FIFO, ...

rivos-eblot replied:
It seems there is no reset handler, so maybe add a comment and/or create a new issue for another PR?

ziuziakowska replied:
https://github.com/lowRISC/qemu/issues/237

rivos-eblot commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,312 @@
+ * THE SOFTWARE.
+ */
+
+/*
```
Maybe move the protocol doc to docs/opentitan/... ?

rivos-eblot commented on hw/opentitan/ot_i2c_host_proxy.c:
```diff
@@ -0,0 +1,312 @@
+            s->parser_state = OT_I2C_CMD_INIT;
+        }
+        break;
+    }
```
Need a defaut case to handle the unexpected cases. Liklely a g_assert_not_reached() since the case should never occur or at least a `qemu_log` or `warn_report`.

rivos-eblot commented on include/hw/opentitan/ot_i2c_transport.h:
```diff
@@ -0,0 +1,37 @@
+#define HW_OPENTITAN_OT_I2C_TRANSPORT_H
+
+#include "qom/object.h"
+#include "hw/sysbus.h"
```
I do not think you need this inclusion here in this simple case: it's all about QEMU objects.

ziuziakowska committed:
[ot] docs: opentitan: document I2C Host Proxy in `i2c_host_proxy.md`

Signed-off-by: Alice Ziuziakowska <a.ziuziakowska@lowrisc.org>

ziuziakowska committed:
[ot] hw/opentitan: ot_i2c: implement `I2CSlaveClass` `send`

Needed for `ot-i2c_host_proxy` to be able to communicate with `ot_i2c`
in target mode.

Signed-off-by: Alice Ziuziakowska <a.ziuziakowska@lowrisc.org>

rivos-eblot commented on docs/opentitan/i2c_transport.md:
```diff
@@ -0,0 +1,39 @@
+# OpenTitan I2C Transport
+
+The `ot-i2c-transport` device provides a way to carry out I2C transactions with QEMU I2C bus devices through an externally driven socket `chardev` interface.
```
I would recommend to keep the wrap width to 100 col max as with existing OT docs in docs/opentitan.

rivos-eblot commented on docs/opentitan/i2c_transport.md:
```diff
@@ -0,0 +1,39 @@
+
+`-chardev socket,id=i2ctp,file=<name>,server=on,wait=off -device ot-i2c-transport,bus=<bus>,chardev=i2ctp`
+
+Where `bus` is any of the OpenTitan Earlgrey I2C buses: `ot-i2c0`, `ot-i2c1`, or `ot-i2c2`.
```
May be rather add a reference to EG and DJ machines, as several OT machines use I2C buses, which may have different naming / count, so it is probably better to move machine specific details to machine documentation.

ziuziakowska replied:
I added a note that these names are for EG. I do not know what the DJ I2C bus name is. 

I think this will need a fix as in EG the devices have a `OT_COMMON_DEV_ID` property that is used to construct the name of the bus in `ot_i2c`. This property is not present in the DJ instantiation.

rivos-eblot commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,296 @@
+static int ot_i2c_transport_can_receive(void *opaque)
+{
+    OtI2CTransportState *s = OT_I2C_TRANSPORT(opaque);
+    int num_free = (int)fifo8_num_free(&s->cmd_bytes_fifo);
```
Not sure to understand the underlying logic:  Is it expected the FIFO free returns a value > INT32_MAX?
Maybe simply cast the result as an `int`?

rivos-eblot commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,296 @@
+static void ot_i2c_transport_receive(void *opaque, const uint8_t *buf, int size)
+{
+    OtI2CTransportState *s = OT_I2C_TRANSPORT(opaque);
+    g_assert(fifo8_num_free(&s->cmd_bytes_fifo) >= size);
```
Maybe add an explicit cast for size

rivos-eblot commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,296 @@
+
+typedef enum {
+    OT_I2C_CMD_INIT,
+    OT_I2C_CMD_ADDR_1,
```
Maybe document the states, as _1 and _2 appears to relate to nibbles?

rivos-eblot commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,296 @@
+{
+    OtI2CTransportState *s = OT_I2C_TRANSPORT(obj);
+    fifo8_create(&s->cmd_bytes_fifo, CMD_FIFO_SIZE);
+    s->parser_state = OT_I2C_CMD_INIT;
```
I think they are duplicates of _reset_enter (and already all zeroed at this point)

rivos-eblot commented on hw/opentitan/Kconfig:
```diff
@@ -75,9 +75,13 @@ config OT_HMAC
 config OT_I2C
     imply I2C_ECHO
     imply DS1338
+    select OT_I2C_TRANSPORT
```
I wonder whether it is a dependency of OT_I2C. It seems that is is more related to general I2C support than specific to OT_I2C block.

Maybe a better fit would be a selection for the OT machines themselves (EarlGrey, Darjeeling)

rivos-eblot commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,296 @@
+            break;
+        }
+        s->byte = (s->byte << 4u) | (uint8_t)nibble;
+        ot_i2c_transport_start_transfer(s);
```
It is expected that on transfer failure state machine resumes execution? I would have expect to move it to an error state which could only be reset once I2C bus is released.
It cannot rely on the remote host which is free to send anything at anytime

rivos-eblot commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,296 @@
+            break;
+        }
+        nibble = g_ascii_xdigit_value(ch);
+        if (nibble < 0) {
```
I think it would be nice to have traces to debug how the remote host is behaving.
Out-of-sync issues may be hard to track down.

rivos-eblot commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,296 @@
+    fifo8_push_all(&s->cmd_bytes_fifo, buf, size);
+
+    /* Drive the command state machine with each byte */
+    while (!fifo8_is_empty(&s->cmd_bytes_fifo)) {
```
It seems a bit fragile: the remote host may pushing data forever, which could deadlock QEMU (I may have missed something though)

I think there should be a mecanism to either pace communication: some monitoring of the I2C frame would be required, where comm would be paced (_i.e._ QEMUTimer) and/or discarded when I2C frames are too long (`R`/`W` without `S` after N bytes?)

ziuziakowska replied:
There does need to be some kind of pacing, especially since software on Ibex might need some time to prepare responses for I2C read transactions. If many bytes are sent at once to read out data, they would all be handled by the I2C subsystem without giving software much opportunity to prepare anything.

But I am not sure how best to do this. I thought for now I should implement it without worrying about that and then implement some form of command buffering on the OpenTitanTool side - to read a byte every say 10ms expecting a correct response by then, or synchronising by another way.

rivos-eblot replied:
I'm not sure to understand what prevents this loop to execute forever when random data (i.e. invalid ones) are pushed to the socket. There should be some kind of control to prevent DoS if something goes wrong on the remote side, whether is it intentional or not. The device should be able to detect this case and drop incoming data and/or close the connection, rather than breaking the whole machine. Maybe I overlooked the code, but I do not see such management. It can be really basic (disable receive, drop the connection, ...), but I think it should exist.

ziuziakowska replied:
I've made an error state now which does not read any bytes from the chardev and is entered if something goes wrong in command parsing, based on `ot_spi_device`. Also removed the FIFO so that one byte at a time is read and processed.

ziuziakowska replied:
Using a timer on the QEMU side to stall processing for a bit so that the vCPU resumes seems to work well enough so I've done that now.

rivos-eblot commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,296 @@
+    }
+}
+
+static int ot_i2c_transport_be_change(void *opaque)
```
Sorry I do not remember the details, but I think you need to also use `qemu_chr_fe_add_watch` (see ot_spi_device, ot_gpio, ... for details) otherwise I remember there may be an issue with re-connections. I forgot the details about this sorry.

rivos-eblot replied:
I do not think it has been implemented. Can you open an issue to implement it with another PR?

AlexJones0 replied:
IMO this should be added in this PR to avoid issues with re-connections - it should be a simple task looking at other CharDev uses, unless it causes further issues to arise.

ziuziakowska replied:
https://github.com/lowRISC/qemu/issues/237

rivos-eblot commented on hw/opentitan/ot_i2c.c:
```diff
@@ -1165,10 +1165,13 @@ static int ot_i2c_target_event(I2CSlave *target, enum i2c_event event)
     }
 
     switch (event) {
+    case I2C_START_SEND:
```
Can you elaborate why SEND_ASYNC could not work with OT_I2C_TRANSPORT?

ziuziakowska replied:
I believe it should, but I do not know what the difference is between `SEND` and `SEND_ASYNC` in QEMU. I chose to use `SEND` for write transactions because it appears that `SEND_ASYNC` is a newer interface which might not be supported by all bus devices.

rivos-eblot commented on hw/opentitan/ot_i2c_host_proxy.c:
```diff
@@ -0,0 +1,103 @@
+/*
```
Could this commit be merged with the actual implementation?

ziuziakowska replied:
As in have one commit for the whole implementation?

rivos-eblot replied:
Yes, it does not seem the first commit provides any feature (?)

rivos-eblot commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -207,15 +212,42 @@ static void ot_i2c_transport_receive(void *opaque, const uint8_t *buf, int size)
-                             ot_i2c_transport_receive, NULL, NULL, state, NULL,
-                             true);
+                             ot_i2c_transport_receive, NULL,
+                             ot_i2c_transport_be_change, state, NULL, true);
```
nit pick: I think it is `BE` for (CharDev)BackEnd (vs. `be`)  in the commit log message.

rivos-eblot commented on docs/opentitan/i2c_transport.md:
```diff
@@ -0,0 +1,39 @@
+
+* For write commands, a written byte will be echoed back on ACK, otherwise `!!` will be sent back on NACK.
+
+* Spaces and newlines in a sent command may appear anywhere except between the pair of hex digits making up a byte, and are ignored. Unexpected characters anywhere in a command will cause the current command to end and command parser restarted.
```
You may want to ignore `\r` and `\t` as additions to `\n`.

rivos-eblot commented on docs/opentitan/i2c_transport.md:
```diff
@@ -0,0 +1,39 @@
+
+* All bytes values are encoded as a pair of case-insensitive ASCII hex digits. For example, decimal 161 is encoded as either `A1` or `a1`.
+
+* End of a command is indicated with an `S`, this ends the transaction.
```
I do not know the capability of the OT I2C target mode capabilities, but how the restart are handled?

_e.g._ in I2C: W A <payload> R A <payload> S
with:
 * W: write
 * R: read
 * A: address
 * S: stop

 BTW not sure to understand how ACK/NACK signalling is handled

ziuziakowska replied:
Looking at it again, I think there is actually a way to perform something of a "repeated-start condition" and change the transfer direction without ever releasing the bus. I will rework the command parser to allow for this.

In QEMU's I2C implementation, `send` to write a byte of data to a device can NACK by returning a non-zero value. `event` can also NACK a start event. All other event and transfer types such as `recv` and `send_async` must succeed. In those cases the special `!!` value is returned instead of any data.

rivos-eblot replied:
Should'nt the protocol over the socket closely stick to the I2C protocol? I'm not sure I have a clear mental picture of the I2C data exchange over the socket vs. a real I2C bus. Maybe it is the case. I think it would be nice to have a short doc (ASCII doc in MD file?) to depict the I2C transport protocol over the socket vs. the physical I2C protocol.

ziuziakowska replied:
I've updated the protocol slightly. Transactions now start with an address byte and then a `R` or `W` for the direction - this is a start condition. After that bytes can be written or read once started, and the transaction can change direction any number of times after that using additional `R`s and `W`s for a repeated-start condition. The transaction is stopped with an `S`.

This is a pretty thin wrapper around the raw I2C functions, but I do not think there is much we can do to make it more robust (e.g, QEMU's I2C implementation does not support multiple controllers at once, and there is nothing stopping anyone with access to the bus from ending any transfer with `i2c_end_transfer` including ones they are not part of, so these behaviours are undefined by us.)

rivos-eblot commented on docs/opentitan/i2c_transport.md:
```diff
@@ -0,0 +1,68 @@
+
+* During a read transaction, a `B` byte can be written to read a byte from the device.
+
+* During a write transaction, written bytes are sent to the device. These will be written back on
```
Not sure to exactly get it:
A real I2C device respond with a ACK/NACK on each written byte, i.e. a NACK may happen at anytime, in which case the master should stop the on-going request. I'm not sure to understand if the protocol fulfills this chronology, as I understand it is more transaction like, but I'm not sure to get it right.
I think it would be nice to have a temporal view of the example, such as
````
BYTE      0        1     2     3
-----|---------|------|-----|-----|
MAS    51W      AA     BB    CC
DEV         51W     AA    BB    !!
````
Not sure my understanding is correct.
Trying to write this make me wonder why we are not sticking to something closer to the actual I2C protocol. 
For example, the first "byte" which requires 3 characters cannot be replied to before the 3 chars, which differs from the payload bytes where the reply can be emitted after 2 chars. I'm wondering if this is not prone to ouf of sync conditions.

ziuziakowska commented:
I've made the wire protocol use raw bytes now instead of hex encoded ones, as I now have a functional implementation to talk to this in `opentitanlib`.

I have also made the device throttle processing using a timer to allow control to be returned to the vCPU so that it can observe the transfers taking place and prepare response bytes. I thought to do this manually in `opentitanlib`, but after implementing it there I think it is better to have it done in the device in QEMU so that it is transparent to all users of it.

rivos-eblot commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,318 @@
+static void ot_i2c_transport_init(Object *obj)
+{
+    OtI2CTransportState *s = OT_I2C_TRANSPORT(obj);
+    s->parser_state = CMD_I2C_IDLE;
```
This is usually part of the reset handler. Do you still encounter issues with the reset functions?

ziuziakowska replied:
I had a look into it, but I think that `I2CSlaveClass`, which this inherits from, is not a "resettable class", so there are no parent reset phases for it.

I do not understand QEMU reset well enough to know how to implement it, and on the `opentitanlib` side we can ensure that commands are always well-formed, but it would be good to implement this at some point.

rivos-eblot replied:
I have a look tomorrow, I guess you can merge the current PR then open a ticket for further improvements?

rivos-eblot commented on hw/opentitan/ot_i2c_host_proxy.c:
```diff
@@ -0,0 +1,318 @@
+        s->parser_state = CMD_I2C_WRITE;
+        break;
+    case CMD_I2C_ERR:
+        break;
```
Maybe add a trace/warning?

rivos-eblot replied:
What about this item? Would it make sense?

rivos-eblot replied:
I guess this is also https://github.com/lowRISC/qemu/issues/237

rivos-eblot reviewed (changes requested):
There is something I deeply do not understand with how the I2C stack is managed, but I do not want to delay this PR any further, as I guess the feature is needed.

I'll try to find some time to dig deeper.

rivos-eblot commented:
nit pick: do not forget to update the format of the PR title

rivos-eblot commented:
I had a very quick look at the I2C target mode documentation, and I think the QEMUTimer is not required, as the target implements clock stretching. I think what is required is that the ACK/NACK (controller to target write) or payload (controller from target read) should only be send when the respective HW FIFO are in the proper state. If the FIFO is full (write) no ACK/NACK should be reported (and obviously not data should be send if the FIFO is empty (read).

It looks like everything can be managed with the communication "transport" protocol, and I'm still not sure it covers all use cases and stick to the actual I2C implementation, such as:
* SCL is only managed by the controller, the target only obey to it: this means the controller (remote peer) should ask for each byte it wants to read. This gives flow control at byte level, exactly like the I2C physical bus
* SCL stretching which is handled by the target would be trivially easy to implement: on poll from the controller to receive one new byte, no reply should be pushed back till SCL stretching is released
* ACK/NACK from the controller to resume a read request should be handled as well (can be mixed with its "poll" query if that makes sense).

If the remote host performs an illegal request (_e.g._ pushing a new byte whereas it has not received a ACK or NACK), the state machine should be moved to ERROR, and only restarted on a new connection for example.

To sum up, I keep thinking it would help to draft some kind of chronogram to compare the I2C physical bus signalling vs. its translation into byte exchange over the comm. socket, which should cover all standard cases.

rivos-eblot commented on docs/opentitan/i2c_transport.md:
```diff
@@ -0,0 +1,47 @@
+the 7-bit target address of the transaction and the read/write bit as the least significant bit
+(0 indicates a write transfer, 1 indicates a read transfer).
+
+- On a repeated start condition, the I2C target address of the first transfer will be used, and
```
I'm not sure to understand how you discriminate signaling (START/STOP/RESTART/ACK/NACK) from payload.

For example in `b's123s123S'`, is the second `s` a restart or the 4th byte of the payload?

ziuziakowska replied:
If you started a write transaction, before each payload byte you send a `W` first to indicate the next byte is payload, or instead of `W` you can use repeated start `s` or stop `S`.

rivos-eblot replied:
ok

ziuziakowska commented:
> I had a very quick look at the I2C target mode documentation, and I think the QEMUTimer is not required, as the target implements clock stretching. I think what is required is that the ACK/NACK (controller to target write) or payload (controller from target read) should only be send when the respective HW FIFO are in the proper state. If the FIFO is full (write) no ACK/NACK should be reported (and obviously not data should be send if the FIFO is empty (read).

This would be ideal, however clock stretching is unimplemented in I2C, as QEMU's I2C model is quite simplistic and I2C operations cannot be deferred until some data ready condition (e.g `I2CSlaveClass` `i2c_recv` must return *something* as the read data byte, and there is no way to express failure).

As a workaround, the timer is used to throttle the I/O to yield control back to the OT I2C, so that it can observe the transactions going on. Otherwise read transfers will all complete immediately on the I/O thread when it decides to process the I2C transfer, and at no point will OT I2C be aware a read transfer is taking place and have no opportunity to fill the TX FIFO.

> SCL is only managed by the controller, the target only obey to it: this means the controller (remote peer) should ask for each byte it wants to read. This gives flow control at byte level, exactly like the I2C physical bus

This is already the case for transfers in both directions - once a read transfer is started, the controller issues a read by sending an `R` byte, and in any transfer the bytes can be sent with any amount of delay.

At some point I would like to implement a better I2C layer to model this better, there are already some wrappers such as `IbexIRQ`. For now I have avoided modifying core QEMU code and tried to work around it.

rivos-eblot commented:
Yes you are right, the I2C API is far too limited, we will need to extend it at some point.

rivos-eblot commented on docs/opentitan/i2c_transport.md:
```diff
@@ -0,0 +1,47 @@
+
+## Protocol
+
+- The protocol over the chardev mirrors the I2C specification closely.
```
There is a discrepancy with the I2C protocol: when reading, the controller is supposed return a ACK/NACK to the slave to inform it whether it wants to keep reading or not.

Note that in I2C spec:
* the START bit is referred as `S`, while
*  the STOP bit is referred to as `P`

I really keep thinking some kind of ASCII art would help understanding the CharDev based protocol vs. the physical protocol

<img width="609" height="640" alt="i2c" src="https://github.com/user-attachments/assets/81b791f3-82bb-49b5-a308-4874dc34253d" />

ziuziakowska replied:
> There is a discrepancy with the I2C protocol: when reading, the controller is supposed return a ACK/NACK to the slave to inform it whether it wants to keep reading or not.

Over the protocol, during a read transfer an `R` byte is sent to read the next byte from the slave, or the transaction could be ended. This would roughly correspond to "ACK and read next byte" and "NACK and stop condition".

> Note that in I2C spec:
> 
>     * the START bit is referred as `S`, while
> 
>     * the STOP bit is referred to as `P`

I think I will rename them then.

rivos-eblot commented on docs/opentitan/i2c_transport.md:
```diff
@@ -0,0 +1,47 @@
+Where `bus` is any of the system's I2C buses. On earlgrey for example, these are `ot-i2c0`,
+`ot-i2c1`, and `ot-i2c2`.
+
+## Protocol
```
I think the protocol is missing a version info. It is really important since the CharDev peer implementation is likely to be part of another repository, which means no way to ensure the same protocol version is used on both sides.

There should be an initial handshake to ensure both peers can exchange knowledgeable data.

ziuziakowska replied:
Is a single byte containing the version before initiating every transaction sufficient?

rivos-eblot replied:
I think it should be enough to perform it when a new connection is open (that is likely translating into another state in your FSM). When a connection is open, wait for some version info (with a prefix as for the other commands, maybe `V`?) and a number. I do not think it is worth having some kind of major/minor version split.

ziuziakowska replied:
Are chardevs in QEMU guaranteed to have at most one active connection? Theoretically `socket` type chardevs could have many active connections at once. Opentitan tooling will always use `pty` type, but other users may use something else so it might be something to worry about.

If at most one active connection can exist, then I can do that with another state in the FSM.

rivos-eblot replied:
Yes, CharDev is either not connected or connected to a single client.
Note: on our side, I think we never use ptys, only unix and TCP sockets.

ziuziakowska replied:
Performing a handshake at the start of a connection in `opentitanlib` is not ideal right now so I've made it be before every transaction, i.e you send the version before the first `S` (you can still do repeated start conditions to change direction) and after every stop condition `P`.

The version is sent as a `i` byte (for I2C) and then a version number (1 for now). If we changed the version handshake to just at the opening of a connection we would increment this but it would still be the first thing sent over the wire.

rivos-eblot replied:
This seems really awkward. Should a protocol being designed around the existing limitation of a single tool? Maybe leave the current opentitantool protocol as it is without a version then so the PR can be merged and initial tests being done, but at some point, I think another protocol will be useful.

AlexJones0 replied:
This isn't a huge limitation on the `opentitanlib` side (it can be worked around), but I think it makes sense to send a version before each transaction - in the same way that I think the SPI Device Chardev protocol currently works with its header that contains a version. This way you are flexible to multiple versions being used via the device, rather than it being set by the first thing that connects to the CharDev and then stuck from then on (due to I2CSlave not inheriting from a resetable class, though that is a separate limitation).

Given that I2C transactions can be arbitrarily long I think it best supports a flexible number of use cases without adding additional maintenance burden or much overhead (my understanding is that I2C transactions can be arbitrarily long, so overheads should be fairly minimal).

If you disagree it should be easy enough to change on `opentitanlib` (in fact, I believe that @ziuziakowska originally had the version at the start before I pointed out that it might cause problems with the current flow due to the use of PTYs).

rivos-eblot replied:
I would have say that I2C packets are generally short, but it really depends on the use case. I was not really concern about the overhead though. On SPI the version field is in each packet because there was some free slot in the 32-bit header, but I do think it would be prefereable to handle it as an initial handshake (such as it is done for the devproxy channel for example). I can't think of a use case where the protocol would change over a session (that is a connection).
Anyway not a big deal, if this is more convenient for now let's keep it the way it is already implemented.

rivos-eblot commented on hw/opentitan/ot_i2c_host_proxy.c:
```diff
@@ -0,0 +1,318 @@
+    s->is_read_command = false;
+    s->address = 0u;
+    s->stall = false;
+    s->stall_timer =
```
Timer is not cancelled on reset, this will bite hard at some point.

AlexJones0 replied:
I think this is not that bad if we just use the timer (and not the additional `stall` attribute) as I suggest in a separate comment, since the stall time is so short and we can then avoid timers causing state change across resets. But yes, maybe this should be documented as part of the lack of resets as discussed in other comments.

rivos-eblot replied:
yes I hadn't seen your other comment.

rivos-eblot reviewed (approved):
Can you open an issue for each open point so this PR can be merged but all the remaining issues are addressed? Please add the link to the comment before resolving it, so it's easier to track the resolution. Thanks.

AlexJones0 commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,318 @@
+/*
+ * OpenTitanTool I2C transport device for OpenTitan
```
Nit: this appears in many places, but I think it would be good if the QEMU model didn't reference OpenTitanTool/"transport"s at all. While it is what we use for testing upstream OpenTitan, it technically is just host-side SW responsible for interacting with OT and could be replaced with any working host SW stack.

If my understanding is correct, any host SW / user can interact with this device via the protocol, and so it doesn't make sense to specifically label "OpenTitanTool". Maybe "I2C host proxy device" or something similar would be more suitable?

ziuziakowska replied:
I think "I2C host proxy device" or "I2C proxy device" could work as a name. "transport" was chosen as the immediate use-case for this was usage in `opentitanlib`, so it made sense to me to call the device after that, but I think now the term is overloaded and confusing.

AlexJones0 commented on include/hw/opentitan/ot_i2c_transport.h:
```diff
@@ -0,0 +1,36 @@
+ * THE SOFTWARE.
+ */
+
+#ifndef HW_OPENTITAN_OT_I2C_TRANSPORT_H
```
Nit: as with above, maybe this might be better as something like `HW_OT_HOST_I2C_PROXY_H`? etc.

rivos-eblot replied:
I think the convention is to follow the exact file path, so that would be `HW_OPENTITAN_OT_I2C_HOST_PROXY_H` for `hw/opentitan/ot_i2c_host_proxy.h` (it's still an OpenTitan implementation, but not tied to OpenTitan tool if I get it right)

AlexJones0 commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,318 @@
+     * to process any I2C transfers.
+     */
+    bool stall;
+    QEMUTimer *stall_timer;
```
I think that this `stall` field is redundant with `timer_pending(&s->stall_timer)` or `timer_expired(&s->stall_timer, qemu_clock_get_ns(OT_VIRTUAL_CLOCK))`, given that all `stall()` does is set the attribute and timer and `unstall()` just unsets the attribute.

AlexJones0 commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,318 @@
+
+static void ot_i2c_transport_put_nack(OtI2CTransportState *s)
+{
+    ot_i2c_transport_put_byte(s, '!');
```
Should these maybe be explicitly `(uint8_t)'!'` etc.?

AlexJones0 commented on hw/opentitan/ot_i2c_host_proxy.c:
```diff
@@ -0,0 +1,318 @@
+            s->parser_state = CMD_I2C_START;
+            break;
+        }
+        s->parser_state = CMD_I2C_ERR;
```
Maybe a warning here? "malformed byte 0x%02x received in idle state"

AlexJones0 commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,318 @@
+
+    /* chardev i2c command parser state */
+    CmdParserState parser_state;
+    bool is_read_command;
```
Nit: this `is_read_command` field feels redundant, since it is given in both the start and repeated start byte. Can we just pass this as an arg to `ot_i2c_transport_start_transfer` instead? Or could it possibly need to be stored for future extensions?

AlexJones0 commented on hw/opentitan/ot_i2c_transport.c:
```diff
@@ -0,0 +1,318 @@
+    OtI2CTransportState *s = OT_I2C_TRANSPORT(opaque);
+
+    qemu_chr_fe_set_handlers(&s->chr, ot_i2c_transport_can_receive,
+                             ot_i2c_transport_receive, NULL,
```
Should this have an event callback to handle the cases where the CharDev is unexpectedly closed (`event == CHR_EVENT_CLOSED`)? Maybe also at some points we should be polling `qemu_chr_fe_backend_connected` and reset/error if appropriate. This might be better documented and left for a future PR however?

AlexJones0 commented on docs/opentitan/i2c_transport.md:
```diff
@@ -0,0 +1,47 @@
+
+`-chardev pty,i2ctp -device ot-i2c-transport,bus=<bus>,chardev=i2ctp`
+
+Where `bus` is any of the system's I2C buses. On earlgrey for example, these are `ot-i2c0`,
```
Minor nit:
```suggestion
Where `bus` is any of the system's I2C buses. On Earl Grey for example, these are `ot-i2c0`,
```

AlexJones0 commented on docs/opentitan/i2c_transport.md:
```diff
@@ -0,0 +1,47 @@
+through an externally driven `chardev` interface.
+
+The device can perform read or write transactions with bus devices and is intended to be used to
+drive the Opentitan I2C device Target mode functionality through `opentitanlib`.
```
Again, maybe remove reference to `opentitanlib` here as any host SW could use this protocol, not just upstream opentitan SW.

AlexJones0 commented on docs/opentitan/i2c_transport.md:
```diff
@@ -0,0 +1,47 @@
+
+- The protocol over the chardev mirrors the I2C specification closely.
+
+- A start and repeated start condition is signalled with a `s` byte, followed by a byte containing
```
Grammar nit (also repeated further down):
```suggestion
- A start and repeated start condition is signalled with an `s` byte, followed by a byte containing
```

AlexJones0 reviewed (commented):
Broadly I agree with previous comments on documenting the deficiencies of QEMU's I2C protocol, so that these issues are captured somewhere for reference in the future if more accurate I2C emulation is desired. I've also left a few separate comments.

ziuziakowska commented:
The device has been renamed to the "I2C Host Proxy" device, as "Transport" was already used to refer to each backend's (e.g QEMU, FPGA, etc.) communication implementations in `opentitanlib`.

I do not think it is possible to implement reset, as the parents class `I2CSlaveClass` does not implement resetting, and the QEMU reset mechanism works by objects inheriting from their parent resettable class and sequencing reset phases around that. A [patch on the mailing list](https://lists.gnu.org/archive/html/qemu-devel/2024-02/msg03324.html) exists but it does not seem wanted.

For now I've made disconnect end any ongoing transaction and reset the state of the parser.

rivos-eblot commented on include/hw/opentitan/ot_i2c_host_proxy.h:
```diff
@@ -0,0 +1,36 @@
+ * THE SOFTWARE.
+ */
+
+#ifndef HW_OPENTITAN_I2C_HOST_PROXY_H
```
Nit pick: To follow the full path name it should be `HW_OPENTITAN_OT_I2C_HOST_PROXY_H` (see for ex. `HW_OPENTITAN_OT_I2C_H` for `hw/opentitan/ot_i2c.h`

rivos-eblot replied:
already done

rivos-eblot commented on include/hw/opentitan/ot_i2c_host_proxy.h:
```diff
@@ -0,0 +1,36 @@
+
+#include "qom/object.h"
+
+#define TYPE_OT_I2C_HOST_PROXY "ot-i2c-host-proxy"
```
nit pick: should be "`ot-i2c_host_proxy`" to match other devices.

rivos-eblot commented on hw/opentitan/ot_i2c_host_proxy.c:
```diff
@@ -0,0 +1,377 @@
+    }
+}
+
+static void ot_i2c_host_proxy_timer(void *opaque)
```
this is the timer callback, so the timer is already "deleted".

rivos-eblot commented on hw/opentitan/ot_i2c_host_proxy.c:
```diff
@@ -0,0 +1,377 @@
+    if (s->parser_state == CMD_I2C_ERR) {
+        return 0;
+    }
+    return timer_pending(s->stall_timer) ? 0 : 1u;
```
1 without u since return value is an int

rivos-eblot commented on hw/opentitan/ot_i2c_host_proxy.c:
```diff
@@ -0,0 +1,377 @@
+{
+    OtI2CHostProxyState *s = OT_I2C_HOST_PROXY(obj);
+    s->parser_state = CMD_I2C_INIT;
+    s->address = 0u;
```
we usually do not zero struct members in init function, since the structure is zeroified by QEMU. It does not hurt though.

rivos-eblot commented on hw/opentitan/ot_i2c_host_proxy.c:
```diff
@@ -0,0 +1,377 @@
+ */
+static void ot_i2c_host_proxy_stall(OtI2CHostProxyState *s)
+{
+    timer_del(s->stall_timer);
```
`timer_mod` does `timer_del`

AlexJones0 commented:
I am not sure why AES tests are failing in regressions recently (maybe an upstream change to the Earlgrey crypto code?), I will have to look into this if I can find time. Regardless I am happy that these CI issues are unrelated so I will merge regardless.
